src/main/java/bio/terra/cli/service/ExternalCredentialsManagerService.java

```diff
@@ -0,0 +1,162 @@
+    }
+  }
+
+  public static ExternalCredentialsManagerService returnFromContext() {
+    return new ExternalCredentialsManagerService(
+        Context.requireUser().getUserAccessToken(), Context.getServer());
+  }
+
+  /**
+   * Gets Ssh key pair from ECM. If one does not exist yet, generate one.
+   *
+   * @param keyPairType git server tied to the key (e.g. GitHub, GitLab, Azure).
+   */
+  public SshKeyPair getSshKeyPair(SshKeyPairType keyPairType) {
+    SshKeyPairApi sshKeyPairApi = new SshKeyPairApi(apiClient);
+    try {
+      return HttpUtils.callWithRetries(
+          () -> sshKeyPairApi.getSshKeyPair(keyPairType),
+          ExternalCredentialsManagerService::isRetryable);
+    } catch (HttpStatusCodeException | InterruptedException e) {
+      if (e instanceof HttpStatusCodeException) {
+        var httpStatusException = (HttpStatusCodeException) e;
+        logErrorMessage(httpStatusException);
+        if (httpStatusException.getStatusCode() == HttpStatus.NOT_FOUND) {
+          return callWithRetries(
+              () -> sshKeyPairApi.generateSshKeyPair(Context.requireUser().getEmail(), keyPairType),
+              "failed to regenerate an ssh key");
+        }
+        throw httpStatusException;
+      }
+      throw new SystemException("Failed to get ssh key", e);
+    }
+  }
+
+  /** Regenerate an SshKey for the user. */
+  public SshKeyPair regenerateSshKeyPair(SshKeyPairType keyPairType) {
+    SshKeyPairApi sshKeyPairApi = new SshKeyPairApi(apiClient);
+    return callWithRetries(
+        () -> sshKeyPairApi.generateSshKeyPair(Context.requireUser().getEmail(), keyPairType),
+        "Failed to regenerate ssh key");
+  }
+
+  /**
+   * Execute a function that includes hitting ECM endpoints. Retry if the function throws an {@link
+   * #isRetryable} exception. If an exception is thrown by the ECM client or the retries, make sure
+   * the HTTP status code and error message are logged.
+   *
+   * @param makeRequest function with a return value
+   * @param errorMsg error message for the the {@link SystemException} that wraps any exceptions
+   *     thrown by the ECM client or the retries
+   */
+  private <T> T callWithRetries(
+      HttpUtils.SupplierWithCheckedException<T, HttpStatusCodeException> makeRequest,
+      String errorMsg) {
+    return handleClientExceptions(
+        () ->
+            HttpUtils.callWithRetries(makeRequest, ExternalCredentialsManagerService::isRetryable),
+        errorMsg);
+  }
+
+  /**
+   * Execute a function that includes hitting ECM endpoints. If an exception is thrown by the ECM
+   * client or the retries, make sure the HTTP status code and error message are logged.
+   *
+   * @param makeRequest function with a return value
+   * @param errorMsg error message for the the {@link SystemException} that wraps any exceptions
+   *     thrown by the ECM client or the retries
+   */
+  private <T> T handleClientExceptions(
```

Comment:
nit: this seems like the kind of function that should be common, and not copied for each service.  but doesn't need to be solved here.